### ISSUE TITLE
How to make OCaml build honour CFLAGS, LDFLAGS et al

### ISSUE THREAD

**Issue by rseichter** [ocaml/ocaml#9191]
Gentoo Linux verifies that all built binaries [honour environment variables](https://devmanual.gentoo.org/ebuild-writing/functions/src_compile/building/) like CFLAGS and LDFLAGS as they are set by our Portage build system. To achieve this, we usually modify the upstream build. For OCaml, I currently use

```shell
sed -i -e 's/\(^OC_LDFLAGS=.*\)/\1 $(LDFLAGS)/' Makefile.config.in
```

before the configure stage. This takes care of the `bin/ocaml*` files, but I am still trying to figure out the best way to ensure that libraries and plugins (`*.so` and `*.cmxs`) also pass Gentoo's QA checks. I tried extending NATDYNLINKOPTS in a similar manner as I did OC_LDFLAGS, but that does not seem to make a difference.

Looking at [Native-code compilation](https://caml.inria.fr/pub/docs/manual-ocaml/native.html), I can't quite figure out the best way to do it, and I would appreciate your suggestions.

**Comment by nojb** [comment]
Duplicate of #8648

**Comment by rseichter** [comment]
@nojb I saw #8648, but I did not consider my issue a duplicate. It would be nice if the OCaml itself took care of things by itself, but for now I would be content with knowing the optimal way my own ebuild should patch things.

**Comment by nojb** [comment]
Sorry, I didn't understand your issue; could you explain for *which* libraries and plugins you would like to set `CFLAGS` and `LDFLAGS` ?

**Comment by rseichter** [comment]
This is the original list of "offenders":

```
QA Notice: Files built without respecting LDFLAGS have been detected
Please include the following list of files in your report:

/usr/bin/ocamlc.opt
/usr/bin/ocamlcmt
/usr/bin/ocamlcp.opt
/usr/bin/ocamldep.opt
/usr/bin/ocamldoc.opt
/usr/bin/ocamllex.opt
/usr/bin/ocamlmklib.opt
/usr/bin/ocamlmktop.opt
/usr/bin/ocamlobjinfo.opt
/usr/bin/ocamlopt.opt
/usr/bin/ocamlprof.opt
/usr/bin/ocamlrun
/usr/bin/ocamlrund
/usr/bin/ocamlruni
/usr/bin/ocamlyacc
/usr/lib64/ocaml/bigarray.cmxs
/usr/lib64/ocaml/libasmrun_shared.so
/usr/lib64/ocaml/libcamlrun_shared.so
/usr/lib64/ocaml/objinfo_helper
/usr/lib64/ocaml/raw_spacetime_lib.cmxs
/usr/lib64/ocaml/str.cmxs
/usr/lib64/ocaml/stublibs/dllcamlstr.so
/usr/lib64/ocaml/stublibs/dllraw_spacetime_lib.so
/usr/lib64/ocaml/stublibs/dllthreads.so
/usr/lib64/ocaml/stublibs/dllunix.so
/usr/lib64/ocaml/unix.cmxs
```

With my change to OC_LDFLAGS, the list has been reduced to the following:

```
/usr/lib64/ocaml/bigarray.cmxs
/usr/lib64/ocaml/libasmrun_shared.so
/usr/lib64/ocaml/libcamlrun_shared.so
/usr/lib64/ocaml/objinfo_helper
/usr/lib64/ocaml/raw_spacetime_lib.cmxs
/usr/lib64/ocaml/str.cmxs
/usr/lib64/ocaml/stublibs/dllcamlstr.so
/usr/lib64/ocaml/stublibs/dllraw_spacetime_lib.so
/usr/lib64/ocaml/stublibs/dllthreads.so
/usr/lib64/ocaml/stublibs/dllunix.so
/usr/lib64/ocaml/unix.cmxs
```

**Comment by rseichter** [comment]
@nojb I now use the following statements to fix the OCaml build:

```shell
sed -i -e 's/\(^MKDLL=.*\)/\1 $(LDFLAGS)/' Makefile.config.in
sed -i -e 's/\(^OC_CFLAGS=.*\)/\1 $(LDFLAGS)/' Makefile.config.in
sed -i -e 's/\(^OC_LDFLAGS=.*\)/\1 $(LDFLAGS)/' Makefile.config.in
```

While this satisfies Gentoo's QA checks, I'd still like to know if this is the best way to do it? By the way, in case you're wondering: Line number 2 which adds LDFLAGS to OC_CFLAGS is indeed deliberate.

**Comment by nojb** [comment]
Hard to give an informed answer as I don't know what Gentoo is passing on the `LDFLAGS` variable, but it looks reasonable to me.

**Comment by rseichter** [comment]
The contents of `LDFLAGS` vary with the different platforms Gentoo supports. Both platform-dependent core requirements and cross-compilation are the main reasons our QA checks are so uncompromising when it comes to environment variables being (dis)respected.

**Comment by gasche** [comment]
I propose a tentative fix in the appropriately-number #10091. If the fix indeed works (it may require a couple iterations), then the changes of @rseichter in https://github.com/ocaml/ocaml/issues/9191#issuecomment-566688620 should not be necessary anymore. (@rseichter, would you by chance be able to check that the fix works for your use-case? I am thinking of cloning the repo and doing a build, then checking the binaries/sharedlibs in the build artifacts.)

**Comment by rseichter** [comment]
@gasche: Thanks Gabriel, but it has been a year and am no longer invested in the OCaml ebuild for Gentoo Linux. I therefore won't be able to verify your proposed fix.
